modules/jfx.incubator.richtext/src/main/java/com/sun/jfx/incubator/scene/control/richtext/CellArrangement.java

```diff
@@ -0,0 +1,389 @@
+                // cells in the top part come after bottom part, and in reverse order
+                return cells.get(bottomCount - ix - 1);
+            }
+        } else if (ix < bottomCount) {
+            // cells in the normal (bottom) part
+            return cells.get(ix);
+        }
+        return null;
+    }
+
+    /** returns a visible cell, or null */
+    public TextCell getVisibleCell(int modelIndex) {
+        int ix = modelIndex - origin.index();
+        if ((ix >= 0) && (ix < visibleCount)) {
+            return cells.get(ix);
+        }
+        return null;
+    }
+
+    /** returns a TextCell from the visible or bottom margin parts, or null */
+    public TextCell getCellAt(int ix) {
+        if (ix < visibleCount) {
+            return cells.get(ix);
+        }
+        return null;
+    }
+
+    public CaretInfo getCaretInfo(Region target, double xoffset, TextPos p) {
+        if (p != null) {
+            int ix = p.index();
+            TextCell cell = getCell(ix);
+            if (cell != null) {
+                int charIndex = p.charIndex();
+                boolean leading = p.isLeading();
+                double dx = -contentPadding.getLeft();
+                PathElement[] path = cell.getCaretShape(target, charIndex, leading, dx, 0.0);
+                if (path == null) {
+                    return null;
+                }
+
+                double lineSpacing = cell.getLineSpacing();
+                return CaretInfo.create(lineSpacing, path);
+            }
+        }
+        return null;
+    }
+
+    public void removeNodesFrom(Pane p) {
+        ObservableList<Node> cs = p.getChildren();
+        for (int i = getVisibleCellCount() - 1; i >= 0; --i) {
+            TextCell cell = cells.get(i);
+            cs.remove(cell);
+        }
+    }
+
+    public void setBottomCount(int ix) {
+        bottomCount = ix;
+    }
+
+    public int bottomCount() {
+        return bottomCount;
+    }
+
+    public void setBottomHeight(double h) {
+        bottomHeight = h;
+    }
+
+    public double bottomHeight() {
+        return bottomHeight;
+    }
+
+    public int topCount() {
+        return cells.size() - bottomCount;
+    }
+
+    public void setTopHeight(double h) {
+        topHeight = h;
+    }
+
+    public double topHeight() {
+        return topHeight;
+    }
+
+    public double averageHeight() {
+        return (topHeight + bottomHeight) / (topCount() + bottomCount);
+    }
+
+    public double estimatedMax() {
+        return (lineCount - topCount() - bottomCount) * averageHeight() + topHeight + bottomHeight;
+    }
+
+    /**
+     * finds a model index of a cell that contains the given localY.
+     * (in vflow frame of reference).
+     * Should not be called with localY outside of this layout sliding window.
+     */
+    private int binarySearch(double localY, int low, int high) {
+        //System.err.println("    binarySearch off=" + off + ", high=" + high + ", low=" + low); // FIX
```

Comment:
Looks like remnants of the debugging session. Do you still need it?

Reply:
yes, indeed it is (to be cleaned up later).
I would encourage the reviewers to focus on the public API rather than the implementation at this stage.